gax/src/test/java/com/google/api/gax/batching/v2/BatchingCallSettingsTest.java

```diff
@@ -0,0 +1,124 @@
+
+  private static final BatchingSettings BATCHING_SETTINGS =
+      BatchingSettings.newBuilder()
+          .setElementCountThreshold(10)
+          .setRequestByteThreshold(20L)
+          .setDelayThreshold(Duration.ofMillis(5))
+          .build();
+  private BatchingCallSettings.Builder<Integer, Integer, LabeledIntList, List<Integer>> builder;
+
+  @Test
+  public void testEmptyBuilder() {
+    builder = BatchingCallSettings.newBuilder(SQUARER_BATCHING_DESC_V2);
+    assertThat(builder.getRetryableCodes()).isEmpty();
+    assertThat(builder.getRetrySettings()).isNotNull();
+
+    builder.setBatchingSettings(BATCHING_SETTINGS);
+    BatchingCallSettings settings = builder.build();
+    assertThat(settings.getBatchingSettings()).isEqualTo(BATCHING_SETTINGS);
+    assertThat(settings.getRetryableCodes()).isEmpty();
+    assertThat(settings.getRetrySettings()).isNotNull();
+  }
+
+  @Test
+  public void testBuilder() {
+    builder = BatchingCallSettings.newBuilder(SQUARER_BATCHING_DESC_V2);
+
+    Set<StatusCode.Code> retryCodes = ImmutableSet.of(StatusCode.Code.UNAVAILABLE);
+    RetrySettings retrySettings = RetrySettings.newBuilder().build();
+    builder
+        .setBatchingSettings(BATCHING_SETTINGS)
+        .setRetryableCodes(retryCodes)
+        .setRetrySettings(retrySettings);
+
+    assertThat(builder.getRetryableCodes()).containsExactly(StatusCode.Code.UNAVAILABLE);
+
+    BatchingCallSettings settings = builder.build();
+    assertThat(settings.getBatchingSettings()).isEqualTo(BATCHING_SETTINGS);
+    assertThat(settings.getRetryableCodes()).isEqualTo(retryCodes);
+    assertThat(settings.getRetrySettings()).isEqualTo(retrySettings);
+  }
+
+  @Test
+  public void testBuilderFromSettings() {
+    builder = BatchingCallSettings.newBuilder(SQUARER_BATCHING_DESC_V2);
+    RetrySettings retrySettings =
+        RetrySettings.newBuilder().setTotalTimeout(Duration.ofMinutes(1)).build();
+    builder
+        .setBatchingSettings(BATCHING_SETTINGS)
+        .setRetryableCodes(StatusCode.Code.UNAVAILABLE, StatusCode.Code.UNAUTHENTICATED)
+        .setRetrySettings(retrySettings);
+
+    BatchingCallSettings settings = builder.build();
+    BatchingCallSettings.Builder newBuilder = settings.toBuilder();
+
+    assertThat(newBuilder.getRetryableCodes())
+        .containsExactly(StatusCode.Code.UNAVAILABLE, StatusCode.Code.UNAUTHENTICATED);
+    assertThat(newBuilder.getRetrySettings()).isEqualTo(retrySettings);
+  }
+
+  @Test
+  public void testMandatorySettings() {
+    Exception actualEx = null;
+    try {
+      BatchingCallSettings.newBuilder(null);
+    } catch (Exception ex) {
+      actualEx = ex;
+    }
+    assertThat(actualEx).isInstanceOf(NullPointerException.class);
+    actualEx = null;
+    try {
+      BatchingCallSettings.newBuilder(SQUARER_BATCHING_DESC_V2).build();
+    } catch (Exception ex) {
+      actualEx = ex;
+    }
+    assertThat(actualEx).isInstanceOf(NullPointerException.class);
```

Comment:
The fact that `BatchingCallSettings.newBuilder(null)` produces `NullPointerException` seems logical (L115), but the fact that `BatchingCallSettings.newBuilder(SQUARER_BATCHING_DESC_V2).build();` produces same - does not. Maybe `IllegalArgumentException` or something like that?

Reply:
good point, I would vote for IllegalStateException

Reply:
Thanks for pointing it out, I have updated BatchingCallSettings to now throw `IllegalStateException`.